lib/MusicBrainz/Server/Data/Utils.pm

```diff
@@ -84,25 +85,37 @@ our @EXPORT_OK = qw(
 
 Readonly my %TYPE_TO_MODEL => map { $_ => $ENTITIES{$_}{model} } grep { $ENTITIES{$_}{model} } keys %ENTITIES;
 
+sub ref_to_type {
+    my $object = shift;
+
+    state %ref_to_type = map {
+        my $model = $TYPE_TO_MODEL{$_};
+        ("MusicBrainz::Server::Entity::$model" => $_)
+    } keys %TYPE_TO_MODEL;
```

Comment:
I assume this not being recalculated every time is a big part of the gain in efficiency? If so, should we be doing the same in `sub model_to_type` which also does a `reverse` every time?

Reply:
That's a big part, and the other is not calling `isa` on every model in the hash until a match is found. Now it just uses a direct lookup. Caching `%MODEL_TO_TYPE` does make sense, I'll push a commit for that.